### PR TITLE
Scripting: Converts casting and def support

### PR DESCRIPTION
**Pull request by stu-elastic**
Painless will cast returned values to a converter
argument type, if necessary.

Painless will also look for a special `convertFromDef`
converter which is called to explicitly handle `def`
conversions.

`convertFromDef` must handle all valid def conversions.

Refs: #59647

### CODE REVIEW

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptClassInfo.java [hunk 1]
```diff
@@ -92,17 +94,33 @@ public ScriptClassInfo(PainlessLookup painlessLookup, Class<?> baseClass) {
...
+        ArrayList<FunctionTable.LocalFunction> converters = new ArrayList<>();
+        FunctionTable.LocalFunction defConverter = null;
         for (java.lang.reflect.Method m : baseClass.getMethods()) {
             if (m.getName().startsWith("convertFrom") &&
                 m.getParameterTypes().length == 1 &&
                 m.getReturnType() == returnType &&
                 Modifier.isStatic(m.getModifiers())) {
 
-                converterSignatures.add(new ConverterSignature(m));
+                if (m.getName().equals("convertFromDef")) {
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptClassInfo.java:105]: TODO: test these failures

**Review by jdconrad — APPROVED** [review]
Nice change!

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptClassInfo.java (hunk 1 above)
Comment (stu-elastic) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/ScriptClassInfo.java:105]: Done

**Comment by javanna** [comment]
@stu-elastic could you add version labels to this PR please?

**Comment by stu-elastic** [comment]
it will be backported to `v7.10.0` with the rest of the new compiler infrastructure.

**Pull request merged**